### PR TITLE
Update root pack

### PR DESCRIPTION
**Pull request by enadeau**
- [x] Add an option to bound the number of rules on requirement placements

- [x] Bound the number of rule on the root placement pack

### CODE REVIEW

File: tilings/strategy_pack.py [hunk 1]
```diff
@@ -276,9 +276,18 @@ def insertion_row_and_col_placements(
...
+        length: int = 3,
+        max_num_req: Optional[int] = 1,
+        max_placement_rules_per_req: Optional[int] = None,
     ) -> "TileScopePack":
-        name = "only_length_{}_root_placements".format(length)
+        if max_num_req is not None:
+            name = "only_length_{length}_{max_num_req}_reqs_root_placements"
+        else:
+            name = "only_length_{length}_root_placements"
+        placment_factory = strat.RequirementPlacementFactory(
```
Comment (jaypantone) [inline tilings/strategy_pack.py:288]: typo

File: tilings/strategy_pack.py [hunk 2]
```diff
@@ -276,9 +276,18 @@ def insertion_row_and_col_placements(
...
+        cls,
+        length: int = 3,
+        max_num_req: Optional[int] = 1,
+        max_placement_rules_per_req: Optional[int] = None,
     ) -> "TileScopePack":
-        name = "only_length_{}_root_placements".format(length)
+        if max_num_req is not None:
+            name = "only_length_{length}_{max_num_req}_reqs_root_placements"
+        else:
+            name = "only_length_{length}_root_placements"
```
Comment (jaypantone) [inline tilings/strategy_pack.py:287]: Do we want `max_placement_rules_per_req` in the name too?

File: tilings/strategy_pack.py (hunk 1 above)
Comment (enadeau) [reply tilings/strategy_pack.py:288]: At least with autocomplete I'm consistent in my typos :)

File: tilings/strategy_pack.py (hunk 2 above)
Comment (enadeau) [reply tilings/strategy_pack.py:287]: I thought about that but it made the name too long in my opinion. We still have all the details of the pack that we ran as we save the JSON of the pack in our database

**Comment by jaypantone** [comment]
I admit I never fully figured out what this pack does, so I'm not the best person to review it.

File: tilings/strategies/requirement_placement.py [hunk 3]
```diff
@@ -453,6 +459,10 @@ def req_indices_and_directions_to_place(
         else:
             all_reqs = tiling.requirements
         for req in all_reqs:
+            if self.max_rules_per_req is not None:
+                num_rule = reduce(lambda prod, req: prod * len(req), req, 1)
```
Comment (christianbean) [inline tilings/strategies/requirement_placement.py:463]: should multiply by 4 for directions

File: tilings/strategy_pack.py [hunk 4]
```diff
@@ -276,9 +276,18 @@ def insertion_row_and_col_placements(
...
+        placement_factory = strat.RequirementPlacementFactory(
+            max_rules_per_req=max_placement_rules_per_req
+        )
```
Comment (christianbean) [inline tilings/strategy_pack.py:290]: now I'm being picky, but shouldn't placement factory variable name just be skippd and it added to expansion strats directly- i don't really care, but it is the way we do it elsewhere

File: tilings/strategies/requirement_placement.py (hunk 3 above)
Comment (enadeau) [reply tilings/strategies/requirement_placement.py:463]: I thought about that, it depends on how we want this flag to play with the `dirs` flag. Maybe we should multiply by the number of directions.

Comment (christianbean) [reply tilings/strategies/requirement_placement.py:463]: Good point, it should multiple len(dirs) as you suggested

File: tilings/strategy_pack.py (hunk 4 above)
Comment (enadeau) [reply tilings/strategy_pack.py:290]: It was ugly when I put that in the strat directly, so I moved it. Not that I really care either.

Comment (christianbean) [reply tilings/strategy_pack.py:290]: ok, don't worry about it, lets fix the other thing and we're good to go!

File: tilings/strategies/requirement_placement.py (hunk 3 above)
Comment (enadeau) [reply tilings/strategies/requirement_placement.py]: Fixed.

**Pull request merged**